src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+        }
+        if (index == SIZE_EMPTY_LIST) {
+            Node<T> newNode = new Node<>(null, value, null);
+            if (size() == SIZE_EMPTY_LIST) {
+                tail = newNode;
+            } else {
+                head.prev = newNode;
+                newNode.next = head;
+            }
+            head = newNode;
+            size++;
+            return;
+        }
+        if (index == size()) {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+            return;
+        }
+        Node<T> currentNode = findByIndex(index);
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+        currentNode.prev.next = newNode;
+        currentNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBoundsException(index);
+        return findByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexOutOfBoundsException(index);
+        T oldValue = findByIndex(index).item;
+        findByIndex(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBoundsException(index);
+        Node<T> currentNode = findByIndex(index);
+        removeNode(currentNode);
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = findByValue(object);
+        if (currentNode == null) {
+            return false;
+        }
+        removeNode(currentNode);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == SIZE_EMPTY_LIST;
+    }
+
+    private Node<T> findByIndex(int index) {
+        Node<T> currentNode = head;
+        for (int i = 1; i <= index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private Node<T> findByValue(T value) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (equals(currentNode.item, value)) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    private void removeNode(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+        size--;
+    }
```

Comment:
i dont sure, but maybe better rename it 

Create a separate method unlink(Node node).

Reply:
ok